cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java

```diff
@@ -298,4 +314,96 @@ public MethodVisitor visitMethod(int access, String name, String desc, String si
+    // We only need to rewrite if listing either HADOOP_CONF_DIR or YARN_CONF_DIR return null.
+    boolean needRewrite = false;
+    for (String env : ImmutableList.of("HADOOP_CONF_DIR", "YARN_CONF_DIR")) {
+      String value = System.getenv(env);
+      if (value != null) {
+        File path = new File(value);
+        if (path.isDirectory() && path.listFiles() == null) {
+          needRewrite = true;
+          break;
+        }
+      }
+    }
+
+    // If rewrite is not needed
+    if (!needRewrite) {
+      return findClass(name);
+    }
+
+    ClassReader cr = new ClassReader(createConfArchive);
+    ClassWriter cw = new ClassWriter(ClassWriter.COMPUTE_MAXS);
+    cr.accept(new ClassVisitor(Opcodes.ASM5, cw) {
+      @Override
+      public MethodVisitor visitMethod(final int access, final String name,
+                                       final String desc, String signature, String[] exceptions) {
+        MethodVisitor mv = super.visitMethod(access, name, desc, signature, exceptions);
+
+        // Only rewrite the createConfArchive method
+        if (!"createConfArchive".equals(name)) {
+          return mv;
+        }
+
+        // Check if it's a recognizable return type.
+        // Spark 1.5+ return type is File
+        boolean isReturnFile = Type.getReturnType(desc).equals(Type.getType(File.class));
+        Type optionType = Type.getObjectType("scala/Option");
+        if (!isReturnFile) {
+          // Spark 1.4 return type is Option<File>
+          if (!Type.getReturnType(desc).equals(optionType)) {
+            // Unknown type. Not going to modify the code.
+            return mv;
+          }
+        }
+
+        // Generate this for Spark 1.5+
+        // return SparkRuntimeUtils.createConfArchive(this.sparkConf, SPARK_CONF_FILE,
+        //                                            LOCALIZED_CONF_DIR, LOCALIZED_CONF_DIR_ZIP);
+        // Generate this for Spark 1.4
+        // return Option.apply(SparkRuntimeUtils.createConfArchive(this.sparkConf, SPARK_CONF_FILE,
+        //                                                         LOCALIZED_CONF_DIR, LOCALIZED_CONF_DIR_ZIP));
+        GeneratorAdapter mg = new GeneratorAdapter(mv, access, name, desc);
+
+        // load this.sparkConf to the stack
+        mg.loadThis();
+        mg.getField(Type.getObjectType("org/apache/spark/deploy/yarn/Client"), "sparkConf", SPARK_CONF_TYPE);
+
+        // push three constants to the stack
+        mg.visitLdcInsn(SPARK_CONF_FILE);
+        mg.visitLdcInsn(LOCALIZED_CONF_DIR);
+        mg.visitLdcInsn(LOCALIZED_CONF_DIR_ZIP);
+
+        // call SparkRuntimeUtils.createConfArchive, return a File and leave it in stack
+        Type stringType = Type.getType(String.class);
+        mg.invokeStatic(Type.getType(SparkRuntimeUtils.class),
+                        new Method("createConfArchive", Type.getType(File.class),
```

Comment:
For this case, no particular reason. However, in this class, we don't use `Methods.getMethod` in other cases due to we cannot refer to Spark/Scala classes with ".class". Also, I would like to minimize dependencies on cdap core classes so that in near future, we can provide better classloader protection for runtime extensions.